### ISSUE TITLE
How do you get current play position programmatically?

### ISSUE THREAD

**Issue by soufriere** [iina/iina#3370]
Hello, I'm a transcriber and I used to use VLC for transcriptions. I have to insert timestamps into my transcript, so I have an applescript that can copy / paste the current VLC play position into my transcript, like this: [00:03:32].

Is there any way to access the play position programmatically in IINA? A terminal command or so? If the play position were in the window title I could access it via applescript, but it doesn't show it in the window title either...

**Comment by low-batt** [comment]
Something like this?:
```bash
low-batt@gag ~$ echo '{ "command": ["get_property", "time-pos"] }' | socat - /tmp/mpv-socket | jq '.data | strftime("%T")'
"00:01:04"
low-batt@gag ~$ 
```

I used [Homebrew](https://brew.sh/) to install [socat](https://formulae.brew.sh/formula/socat) and [jq](https://formulae.brew.sh/formula/jq).

Then I brought up IINA and selected Preferences / Advanced and enabled advanced settings. Under "Additional mpv options" I clicked on `+` and added:
| Name | Value |
| --- | --- |
|input-ipc-server | /tmp/mpv-socket |

I then restarted IINA for the changes to take effect.

IINA is a graphical user interface for the [mpv media player](https://mpv.io/). That player is designed for command line use and has no real GUI, only a small controller on top of the video for basic control. IINA provides a full featured GUI for mpv. Many mpv features are available in IINA. From the [mpv manual](https://mpv.io/manual/stable/#input):

> **--input-ipc-server=\<filename>**
>
>  Enable the IPC support and create the listening socket at the given path.

This allows us to use mpv's [JSON IPC API](https://mpv.io/manual/stable/#json-ipc).  See that section of the manual for examples. This sends a command to get the current position in seconds:

```bash
low-batt@gag ~$ echo '{ "command": ["get_property", "time-pos"] }' | socat - /tmp/mpv-socket
{"data":64.254000,"request_id":0,"error":"success"}
low-batt@gag ~$ 
```

Then [jq](https://stedolan.github.io/jq/) is used to parse the JSON, extract the value of the data key and convert the seconds to a time string.

**Comment by ELgit** [comment]
Wouldn't it be easier to just have playback time as an optional sticky display in IINA so when someone is transcribing or taking notes on the video in a separate app all they have to do is glance at the playback time and write it down?

Given that's not an option/feature, I tried the instructions above.

There are some assumed steps that I (a non-programmer) am not aware of so it's not working for me. 
This is what I did:
- installed home-brew, socat and jq using the terminal commands from the links/pages provided. 
- added the advanced settings per instructions and restarted IINA
- read through the "--input-ipc-server=<filename>" info -- is that a command I invoke somewhere? if so that's not stated
- I don't understand how to use JSON IPC API to send the command
- I don't understand how to tell JSON IPC API to send the command
- after the command is sent, I don't understand how to tell JQ to parse, extract, covert to time string
- I don't understand how to do a keyboard shortcut to invoke everything above and past the time string into the word document I'm using to transcribe the video.

Any help greatly appreciated.

**Comment by low-batt** [comment]
@ELgit Yes, my above instructions could use some more details. Don't worry, we will get this to work for you.

The following is a **partial** answer. I need to get this working using [Automator](https://support.apple.com/guide/automator/welcome/mac). Using Automator we should be able to automate this so that one key press inserts the time. Still working on that. Ran out of time today. For now lets get this working for you from terminal.

> Wouldn't it be easier to just have playback time as an optional sticky display in IINA so when someone is transcribing or taking notes on the video in a separate app all they have to do is glance at the playback time and write it down?

Yes. I'm sure we could come up with an easier way to do this with a new feature. But the question posed by this issue was not about that. So the answer given addresses how to do this with IINA as it is today. A request for a new IINA feature to make this easier should be entered as a new feature request issue. There are plans to add AppleScript support. That might make this easier.

I should have included a picture showing how to use `input-ipc-server`. Bring up IINA/Preferences. Click on the `Advanced` tab. Slide the `Enable advanced settings` toggle button to be on (blue). In the `Additional mpv options` table click on the `+` button. That will add an entry to the table with the values `name` and `value`. Click on `name` and change that to `input-ipc-server`. Click on `value` and change that to `/tmp/mpv-socket`. When finished the preferences window should look like:

![issue-3370-prefereneces](https://user-images.githubusercontent.com/86170219/174878051-dd8c05cb-2f2c-4e4d-a8be-36afad277078.jpg)

Then restart IINA.

This setting causes IINA to open a socket for [inter-process communication](https://en.wikipedia.org/wiki/Inter-process_communication) (IPC) at `/tmp/mpv-socket` and listen for commands sent to that socket. This provides a way to control IINA from terminal.

Once IINA is listening for commands then while playing a movie you should be able to paste the following command sequence into terminal:

```text
echo '{ "command": ["get_property", "time-pos"] }' | socat - /tmp/mpv-socket | jq -r '.data | strftime("%T")' | pbcopy
```

After running that command pasting into a TextEdit window added this to the document:
```text
00:00:24
```

To update the pasteboard with the  press up arrow in terminal to recall the previous command and then press return again.

The technologies involved in the command sequence:
- [pipeline](https://www.gnu.org/software/bash/manual/html_node/Pipelines.html#Pipelines) A sequence of commands connected by the pipe operator (`|`)
- [JSON](https://www.json.org/json-en.html) A standard for passing data as name/value pairs
- [echo](https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html#index-echo) Outputs a JSON formatted string to socat
- [socat](https://www.unix.com/man-page/centos/1/socat/) Writes to and reads from IINA
- [jq](https://stedolan.github.io/jq/) Parses the JSON returned by IINA and extracts and formats the time
- [pbcopy](https://www.unix.com/man-page/osx/1/pbcopy/) Copies the text to the pasteboard

The next step is to get something like this running under Automator and assign a shortcut key. I will investigate that.

**Comment by lhc70000** [comment]
This can be easily done under the [plugin system](https://github.com/iina/iina/pull/3762), which should be available in 1.4.0. We can release a beta version as soon as it's merged.
